src/protocol/boolean/random_bits.rs

```diff
@@ -0,0 +1,129 @@
+use crate::error::BoxError;
+use crate::ff::{Field, Int};
+use crate::protocol::boolean::BitOpStep;
+use crate::protocol::modulus_conversion::convert_shares::{ConvertShares, XorShares};
+use crate::protocol::{context::ProtocolContext, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+
+/// This protocol generates a sequence of uniformly random sharing of bits in `F_p`.
+/// Adding these 3-way secret-sharing will yield the secret `r_i ∈ {0,1}`.
+pub struct RandomBits {}
+
+impl RandomBits {
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<(Vec<Replicated<F>>, Replicated<F>), BoxError> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
+
+        // Generate a replicated random field values. We'll use these fields as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let r_p: Replicated<F> = ctx.prss().generate_replicated(record_id);
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let futures = (0..l).map(|i| {
+            // again, we don't expect our prime field to be > 2^64
+            #[allow(clippy::cast_possible_truncation)]
+            let xor_shares = XorShares {
+                num_bits: l,
+                packed_bits_left: r_p.left().as_u128() as u64,
+                packed_bits_right: r_p.right().as_u128() as u64,
+            };
+            let c = ctx.narrow(&BitOpStep::Step(i as usize));
+            async move {
+                ConvertShares::new(xor_shares)
+                    .execute_one_bit(c, record_id, i as u8)
+                    .await
+            }
+        });
+        let r_b = try_join_all(futures).await?;
+
+        Ok((r_b, r_p))
```

Comment:
Don't bother returning r_p. This has no meaning. It's not an equivalent secret-sharing of the same value. You're having to use a special XOR reconstruction in your test because it's not. This is what I demonstrated in that Google sheet from a few days ago, that this approach does NOT implement the SOLVED-BITs protocol. This is a perfectly fine way of generating a sequence of shared bits, which is like steps 1 and 2 of the SOLVED-BITs protocol, but it's not equivalent to the rest - we will still need to do all of that work. Specifically it's step 6 of that protocol which you need to compute the sharing of the random number in the large prime field.